cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -310,108 +318,127 @@ class GcsFileSystem::Impl {
+      return client_.GetObjectMetadata(bucket, canonical);
+    }
+    return object;
   }
 
-  google::cloud::Status CreateDirMarkerRecursive(const std::string& bucket,
-                                                 const std::string& object) {
-    using GcsCode = google::cloud::StatusCode;
+  static Status NotDirectoryError(const gcs::ObjectMetadata& o) {
+    return Status::IOError(
+        "Cannot create directory, it conflicts with an existing file '",
+        internal::ConcatAbstractPath(o.bucket(), o.name()), "'");
+  }
+
+  Status CreateDirMarkerRecursive(const std::string& bucket, const std::string& name) {
     auto get_parent = [](std::string const& path) {
       return std::move(internal::GetAbstractPathParent(path).first);
     };
-    // Maybe counterintuitively we create the markers from the most nested and up. Because
-    // GCS does not have directories creating `a/b/c` will succeed, even if `a/` or `a/b/`
-    // does not exist.  In the common case, where `a/b/` may already exist, it is more
-    // efficient to just create `a/b/c/` and then find out that `a/b/` was already there.
-    // In the case where none exists, it does not matter which order we follow.
-    auto status = CreateDirMarker(bucket, object);
-    if (status.code() == GcsCode::kAlreadyExists) return {};
-    if (status.code() == GcsCode::kNotFound) {
-      // Missing bucket, create it first ...
-      status = client_.CreateBucket(bucket, gcs::BucketMetadata()).status();
-      if (status.code() != GcsCode::kOk && status.code() != GcsCode::kAlreadyExists) {
-        return status;
+    // Find the list of missing parents. In the process we discover if any elements in
+    // the path are files, this is unavoidable as GCS does not really have directories.
+    std::vector<std::string> missing_parents;
+    auto dir = name;
+    for (; !dir.empty(); dir = get_parent(dir)) {
+      auto o = client_.GetObjectMetadata(bucket, dir);
+      if (o) {
+        if (IsDirectory(*o)) break;
+        return NotDirectoryError(*o);
       }
+      missing_parents.push_back(dir);
     }
-
-    for (auto parent = get_parent(object); !parent.empty(); parent = get_parent(parent)) {
-      status = CreateDirMarker(bucket, parent);
-      if (status.code() == GcsCode::kAlreadyExists) {
-        break;
+    if (dir.empty()) {
+      // We could not find any of the parent directories in the bucket, the last step is
+      // to find out if the bucket exists, and if necessary, create it
+      google::cloud::StatusOr<gcs::BucketMetadata> b = client_.GetBucketMetadata(bucket);
+      if (!b) {
+        if (b.status().code() == GcsCode::kNotFound) {
+          b = client_.CreateBucket(bucket, gcs::BucketMetadata());
+        }
+        if (!b) return internal::ToArrowStatus(b.status());
       }
-      if (!status.ok()) {
-        return status;
+    }
+    for (auto d = missing_parents.rbegin(); d != missing_parents.rend(); ++d) {
```

Comment:
For the record, is it useful to iterate these in reverse order? Is it to avoid creating `a/b` if `a` exists as a file?

Reply:
Yes. Added a comment to explain that.

Reply:
Is the comment a bit off? It seems `missing_parents` is created in reverse depth order (from child to parent), so iterating it in reverse order visits the ancestors first and the descendents last?